plugins/linux-capture/pipewire.c

```diff
@@ -366,6 +375,157 @@ static void swap_texture_red_blue(gs_texture_t *texture)
+
+	/* modifier */
+	if (modifier_count > 0) {
+		/* build an enumeration of modifiers */
+		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_modifier,
+				     SPA_POD_PROP_FLAG_MANDATORY |
+					     SPA_POD_PROP_FLAG_DONT_FIXATE);
+		spa_pod_builder_push_choice(b, &f[1], SPA_CHOICE_Enum, 0);
+		/* modifiers from  an array */
+		for (uint32_t i = 0; i < modifier_count; i++) {
+			uint64_t modifier = modifiers[i];
+			spa_pod_builder_long(b, modifier);
+			if (i == 0)
+				spa_pod_builder_long(b, modifier);
+		}
+		spa_pod_builder_pop(b, &f[1]);
+	}
+	/* add size and framerate ranges */
+	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_size,
+			    SPA_POD_CHOICE_RANGE_Rectangle(
+				    &SPA_RECTANGLE(320, 240), // Arbitrary
+				    &SPA_RECTANGLE(1, 1),
+				    &SPA_RECTANGLE(8192, 4320)),
+			    SPA_FORMAT_VIDEO_framerate,
+			    SPA_POD_CHOICE_RANGE_Fraction(
+				    &SPA_FRACTION(ovi->fps_num, ovi->fps_den),
+				    &SPA_FRACTION(0, 1), &SPA_FRACTION(360, 1)),
+			    0);
+	return spa_pod_builder_pop(b, &f[0]);
+}
+
+static uint32_t build_format_params(obs_pipewire_data *obs_pw,
+				    struct spa_pod_builder *pod_builder,
+				    const struct spa_pod ***param_list)
+{
+	uint32_t n_params = 0;
+
+	const struct spa_pod **params;
+	params =
+		bzalloc(2 * obs_pw->format_info.num * sizeof(struct spa_pod *));
+
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		if (obs_pw->format_info.array[i].modifiers.num == 0 ||
+		    !check_pw_version(&obs_pw->server_version, 0, 3, 33)) {
+			continue;
+		}
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format,
+			obs_pw->format_info.array[i].modifiers.array,
+			obs_pw->format_info.array[i].modifiers.num);
+	}
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format, NULL, 0);
+	}
+	*param_list = params;
+	return n_params;
+}
+
+static bool drm_format_available(uint32_t drm_format, uint32_t *drm_formats,
+				 size_t n_drm_formats)
+{
+	for (size_t j = 0; j < n_drm_formats; j++) {
+		if (drm_format == drm_formats[j]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static void init_format_info(obs_pipewire_data *obs_pw)
+{
+	da_init(obs_pw->format_info);
+
+	uint32_t formats[] = {
+		SPA_VIDEO_FORMAT_BGRA,
+		SPA_VIDEO_FORMAT_RGBA,
+		SPA_VIDEO_FORMAT_BGRx,
+		SPA_VIDEO_FORMAT_RGBx,
+	};
+
+	size_t n_formats = sizeof(formats) / sizeof(formats[0]);
+
+	obs_enter_graphics();
+
+	enum gs_dmabuf_flags dmabuf_flags;
+	uint32_t *drm_formats = NULL;
+	size_t n_drm_formats;
+	bool capabilities_queried = gs_query_dmabuf_capabilities(
+		&dmabuf_flags, &drm_formats, &n_drm_formats);
+	for (size_t i = 0; i < n_formats; i++) {
+		struct format_info *info =
+			da_push_back_new(obs_pw->format_info);
+		da_init(info->modifiers);
+		info->spa_format = formats[i];
+		spa_pixel_format_to_drm_format(formats[i], &info->drm_format);
```

Comment:
It just occurred to me that this function returns `bool`, but the value is ignored now. Is this intended? It seems to me that you could've written this as:

```c
struct format_info *info;
uint32_t drm_format;

if (!spa_pixel_format_to_drm_format(formats[i], &drm_format))
	continue;

...
```

(Consequently, the `formats` array could be removed.)

Reply:
Ah, I think I got it upside down. You iterate over the `formats` array, not the `drm_formats` one. So let me suggest the following: move the `formats` array to right above the `spa_pixel_format_to_drm_format()` function, make it static.

Reply:
two nits about this change: When building the EnumParams modifiers.num is used to check if some are available. Since `da_push_back_new` initalizes everything with 0 we should be fine without initializing the modifier darray. And freeing a {0} darray should also be fine. 
So everything will work fine, but the reason is implicit.